circuit/algorithms/src/keccak/hash.rs

```diff
@@ -0,0 +1,454 @@
+    /// where `b` is the width of the permutation, and `c` is the capacity.
+    fn pad_sha3(input: &[Boolean<E>], bitrate: usize) -> Vec<Vec<Boolean<E>>> {
+        debug_assert!(bitrate > 1, "The bitrate must be greater than 1");
+
+        // Resize the input to a multiple of 8.
+        let mut padded_input = input.to_vec();
+        padded_input.resize((input.len() + 7) / 8 * 8, Boolean::constant(false));
+
+        // Step 1: Append the "0x06" byte to the message.
+        padded_input.push(Boolean::constant(false));
+        padded_input.push(Boolean::constant(true));
+        padded_input.push(Boolean::constant(true));
+        padded_input.push(Boolean::constant(false));
+
+        // Step 2: Append "0" bits until the length of the message is congruent to r-1 mod r.
+        while (padded_input.len() % bitrate) != (bitrate - 1) {
+            padded_input.push(Boolean::constant(false));
+        }
+
+        // Step 3: Append the bit "1" to the message.
+        padded_input.push(Boolean::constant(true));
+
+        // Construct the padded blocks.
+        let mut result = Vec::new();
+        for block in padded_input.chunks(bitrate) {
+            result.push(block.to_vec());
+        }
+        result
+    }
+
+    /// The permutation `f` is a function that takes a fixed-length input and produces a fixed-length output,
+    /// defined as `f = Keccak-f[b]`, where `b := 25 * 2^l` is the width of the permutation,
+    /// and `l` is the log width of the permutation.
+    ///
+    /// The round function `R` is applied `12 + 2l` times, where `l` is the log width of the permutation.
+    fn permutation_f<const WIDTH: usize, const NUM_ROUNDS: usize>(
+        input: Vec<Boolean<E>>,
+        round_constants: &[U64<E>],
+        rotl: &[usize],
+    ) -> Vec<Boolean<E>> {
+        debug_assert_eq!(input.len(), WIDTH, "The input vector must have {WIDTH} bits");
+        debug_assert_eq!(
+            round_constants.len(),
+            NUM_ROUNDS,
+            "The round constants vector must have {NUM_ROUNDS} elements"
+        );
+
+        // Partition the input into 64-bit chunks.
+        let mut a = input.chunks(64).map(U64::from_bits_le).collect::<Vec<_>>();
+        // Permute the input.
+        for round_constant in round_constants.iter().take(NUM_ROUNDS) {
+            a = Self::round(a, round_constant, rotl);
+        }
+        // Return the permuted input.
+        a.into_iter().flat_map(|e| e.to_bits_le()).collect()
+    }
+
+    /// The round function `R` is defined as follows:
+    /// ```text
+    /// R = ι ◦ χ ◦ π ◦ ρ ◦ θ
+    /// ```
+    /// where `◦` denotes function composition.
+    fn round(a: Vec<U64<E>>, round_constant: &U64<E>, rotl: &[usize]) -> Vec<U64<E>> {
+        debug_assert_eq!(a.len(), MODULO * MODULO, "The input vector 'a' must have {} elements", MODULO * MODULO);
+
+        const WEIGHT: usize = MODULO - 1;
```

Comment:
this const can be removed, as all of the vecs in this method require the capacity of either `MODULO` or `MODULO * MODULO`

Reply:
abfb6d408